### PR TITLE
[GEOS-9327]: ncWMS getTimeSeries supporting time list and interval wi…

### PR DESCRIPTION
**Pull request by dromagnoli**
…th period

<Include a few sentences describing the overall goals for this Pull Request>

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [x] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [x] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by travis-ci after opening this PR)
- [x] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.

### CODE REVIEW

File: src/main/src/main/java/org/geoserver/catalog/util/ReaderDimensionsAccessor.java [hunk 1]
```diff
@@ -194,7 +194,7 @@ public boolean hasTime() throws IOException {
         // if we got here, the optimization did not work, do the normal path
         if (result == null) {
             result = new TreeSet<Object>();
-            TreeSet<Object> fullDomain = getElevationDomain();
+            TreeSet<Object> fullDomain = getTimeDomain();
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/catalog/util/ReaderDimensionsAccessor.java]: Ouch good find...

File: src/community/ncwms/src/main/java/org/geoserver/wms/ncwms/NcWmsService.java [hunk 2]
```diff
@@ -70,6 +75,58 @@ public void checkTimeout() {
...
+                    throws IOException {
+                TreeSet<Date> availableDates = new TreeSet<Date>();
+                final boolean isRange = times.get(0) instanceof DateRange;
+                for (Object time : times) {
+                    TreeSet<Object> foundTimes =
+                            wms.queryCoverageTimes(
+                                    coverage, getAsRange(time, isRange), Query.DEFAULT_MAX);
+                    foundTimes
+                            .stream()
+                            .sequential()
```
Comment (aaime) [inline src/community/ncwms/src/main/java/org/geoserver/wms/ncwms/NcWmsService.java]: Give you're populating a treeset, does it really have to be sequential?

File: src/community/ncwms/src/main/java/org/geoserver/wms/ncwms/NcWmsService.java [hunk 3]
```diff
@@ -70,6 +75,58 @@ public void checkTimeout() {
...
+                            .forEach(
+                                    d -> {
+                                        Date date = (Date) d;
+                                        if (!availableDates.contains(date)) {
```
Comment (aaime) [inline src/community/ncwms/src/main/java/org/geoserver/wms/ncwms/NcWmsService.java]: It's a set, no need for explicit checks, just add

File: src/wms/src/main/java/org/geoserver/wms/WMS.java [hunk 4]
```diff
@@ -1274,6 +1274,33 @@ public boolean isQueryable(LayerGroupInfo layerGroup) {
...
+    public TreeSet<Date> queryCoverageNearestMatchTimes(
+            CoverageInfo coverage, List<Object> queryRanges) throws IOException {
+        DimensionInfo time = coverage.getMetadata().get(ResourceInfo.TIME, DimensionInfo.class);
+        final TreeSet<Date> foundDates = new TreeSet<Date>();
+        if (time == null || !time.isEnabled()) {
+            throw new ServiceException(
+                    "Coverage " + coverage.prefixedName() + " does not have time support enabled");
+        }
+
+        List<Object> dates = getNearestMatches(coverage, time, queryRanges, ResourceInfo.TIME);
```
Comment (aaime) [inline src/wms/src/main/java/org/geoserver/wms/WMS.java]: Same comments as in the enum above... come to think of it, the code is pretty similar, is there duplication here?

File: src/community/ncwms/src/main/java/org/geoserver/wms/ncwms/NcWmsService.java (hunk 2 above)
Comment (dromagnoli) [reply src/community/ncwms/src/main/java/org/geoserver/wms/ncwms/NcWmsService.java]: nope, I was collecting to a List at the beginning.
I'm going to fix, thx

File: src/community/ncwms/src/main/java/org/geoserver/wms/ncwms/NcWmsService.java (hunk 3 above)
Comment (dromagnoli) [reply src/community/ncwms/src/main/java/org/geoserver/wms/ncwms/NcWmsService.java]: Right.

File: src/wms/src/main/java/org/geoserver/wms/WMS.java (hunk 4 above)
Comment (dromagnoli) [reply src/wms/src/main/java/org/geoserver/wms/WMS.java]: Actually not. The available WMS.queryCoverageTimes was getting one dateRange at time so the "aggregation" is made outside by the enum method doing the query calls.
The new WMS.queryNearestCoverageTimes accept multiple times (a List) so the aggregation is made inside.
I'm fixing the sequential + containment checks here too.

**Comment by dromagnoli** [comment]
I have handled the improvements suggested in the comments. I'm going to merge once the tests are passing.

**Comment by dromagnoli** [comment]
Merging it. (The appveyor failure was the usual random failure on GWC)

**Pull request merged**